src/Orleans.Core/Networking/ConnectionManager.cs

```diff
@@ -53,158 +55,162 @@ public int ConnectionCount
-                while (!acquiredConnectionLock)
+                lock (this.lockObj)
                 {
-                    entry = GetOrCreateEntry(endpoint, ref acquiredConnectionLock);
-
-                    if (entry.Connections.Length >= this.connectionOptions.ConnectionsPerEndpoint)
+                    if (entry.PendingConnection != null
+                        && entry.Connections.Length < this.connectionOptions.ConnectionsPerEndpoint)
                     {
-                        result = entry.Connections;
-                        break;
+                        // Start a new connection attempt in the background.
+                        this.GetConnectionAsync(endpoint).Ignore();
                     }
-
-                    ThrowIfRecentFailure(endpoint, entry);
-
-                    // Wait a short time before reattempting to acquire the lock
-                    await Task.Delay(10);
-                }
-
-                // Attempt to connect.
-                Connection connection = default;
-                try
-                {
-                    connection = await this.ConnectAsync(endpoint);
-                    entry = OnConnectedInternal(endpoint, connection);
-                    result = entry.Connections;
-                }
-                catch (Exception exception)
-                {
-                    OnConnectionFailed(endpoint, DateTime.UtcNow);
-                    throw new ConnectionFailedException(
-                        $"Unable to connect to endpoint {endpoint}. See {nameof(exception.InnerException)}", exception);
                 }
             }
-            finally
-            {
-                if (acquiredConnectionLock) entry.ReleaseLock();
-            }
 
-            nextConnection = (nextConnection + 1) % result.Length;
-            return result[nextConnection];
+            return new ValueTask<Connection>(connection);
         }
 
-        private void OnConnectionFailed(SiloAddress address, DateTime lastFailure)
+        private async Task<Connection> GetConnectionAsync(SiloAddress endpoint)
         {
-            bool acquiredConnectionLock = false;
-            ConnectionEntry entry = default;
-            lock (this.collectionLock)
+            while (true)
             {
-                try
+                var entry = this.GetOrCreateEntry(endpoint);
+                if (entry.Connections.Length >= this.connectionOptions.ConnectionsPerEndpoint
+                    && entry.TryGetNextConnection(out var connection))
+                {
+                    return connection;
+                }
+
+                // Wait for the pending attempt to complete
+                var pendingTask = entry.PendingConnection;
+                if (pendingTask != null && !pendingTask.IsCompleted)
                 {
-                    entry = this.GetOrCreateEntry(address, ref acquiredConnectionLock);
+                    await pendingTask.ConfigureAwait(false);
+                }
```

Comment:
Previously there was a kind of 'async spinlock' whereby only one connection attempt for a given endpoint could execute at a time and the others would await a `Task.Delay(...)` before re-checking. Instead we now share a `Task` and other connection attempts await that.